[flang][debug] Support assumed size arrays.

abidh opened the pull request:
Here we don't know the size of last dimension and use a null subrange for that. User will have to provide the dimension when evaluating variable of this type. The debugging looks like as follows:

subroutine fn(a1, a2)
 integer  a1(5, *), a2(*)
...
end

(gdb) p a1(1,2)
$2 = 2
(gdb) p a2(3)
$3 = 3
(gdb) ptype a1
type = integer (5,*)
(gdb) ptype a2
type = integer (*)

abidh committed:
[flang][debug] Support assumed size arrays.

Here we don't know the size of last dimension and use a null subrange
for that dimension. User will have to provide the dimension when
evaluating variable of this type. The debugging looks like as follows:

subroutine fn(a1, a2)
 integer  a1(5, *), a2(*)
...
end

(gdb) p a1(1,2)
$2 = 2
(gdb) p a2(3)
$3 = 3
(gdb) ptype a1
type = integer (5,*)
(gdb) ptype a2
type = integer (*)

tblah commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -155,28 +155,31 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertSequenceType(
-    elements.push_back(subrangeTy);
+    if (dim == seqTy.getUnknownExtent()) {
+      auto subrangeTy = mlir::LLVM::DISubrangeAttr::get(
+          context, nullptr, nullptr, nullptr, nullptr);
```
nit: comments with `/*argument_name=*/nullptr`

abidh replied:
Thanks for catching this. Fixed now.

tblah reviewed (approved):
LGTM

abidh committed:
Handle review comments.

Added comments on constant value parameters.

jeanPerier commented on flang/lib/Optimizer/Transforms/DebugTypeGenerator.cpp:
```diff
@@ -155,28 +155,32 @@ mlir::LLVM::DITypeAttr DebugTypeGenerator::convertSequenceType(
-    auto subrangeTy = mlir::LLVM::DISubrangeAttr::get(
-        context, countAttr, lowerAttr, nullptr, nullptr);
-    elements.push_back(subrangeTy);
+    if (dim == seqTy.getUnknownExtent()) {
+      auto subrangeTy = mlir::LLVM::DISubrangeAttr::get(
+          context, /*count=*/nullptr, /*lowerBound=*/nullptr,
+          /*upperBound*/ nullptr, /*stride*/ nullptr);
+      elements.push_back(subrangeTy);
```
Will it be needed to make a difference between assumed-size and explicit shape with non constant extent here?

```
subroutine sub1(x, n)
 real :: x(n)
end subroutine
```

and:
```
subroutine sub1(x)
 real :: x(*)
end subroutine
```

abidh replied:
Thanks for catching this. We need to make a different between these 2 cases. For the 2nd case, we can generate a DISubrange pointing to a variable. But that will require DeclOp to be propagated to DebugTypeConvertor. This information is not available in Type only.

For this PR, I have added a FIXME and a testcase with variable extent arrays that is currently xfailed. I will also open an issue to keep track of this once this PR is merged.

tblah replied:
Won't reading from the descriptor at runtime work just as well in this case? Does the debug output look different for reading from the variable?

abidh replied:
As I see there is no descriptor in this case. The type of a variable I get is `!fir.array<?xf32>`

jeanPerier replied:
Ok, we should probably add some attribute to tag assumed-size array xxx.declare (we can currently infer that looking at the last extent value that should be a -1 constant, but chain lookups are a bit weak IMHO, better to set an attribute.

> As I see there is no descriptor in this case.

To be more specific @tblah, there was one in HLFIR, but it is "unlink" when creating the fir.declare that has a single result. , and it may have been eliminated if unused in FIR.

abidh committed:
Add comments about arrays with non const extent.

Currently, the arrays with non constant extent are being handled like
assumed size arrays. Added a FIXME to mention that there is a better
way to handle them. Also added a testcase that is currently xfailed.
